colflow,colexec: fixes around cancellation

yuzefovich opened the pull request:
**colflow: fix cancellation of the outbox**

Previously, we were passing in the cancellation function of the flow's
context to each outbox which would result in too early of a cancellation
whenever any outboxes encounters a stream error. This is now fixed by
deriving a separate child context with separate cancellation function
for each outbox and only canceling the flow's context when the last
outbox exits.

Release note (bug fix): CockroachDB could previously encounter benign
internal "context canceled" errors when queries were executed by the
vectorized engine.

**colexec: swallow "query canceled" in synchronizer with internal cancellation**

Parallel unordered synchronizer shuts down all of its input subtrees by
canceling the context which those inputs are listening on. This results
in an "internal cancellation", and the error handling is different in
such case in the synchronizer - it needs to swallow "context canceled"
errors because, well, it was the one that canceled it, so such error is
expected part of the "graceful" shutdown. However, we also need to
swallow "query canceled" errors because `CancelChecker` operator might
be listening on `ctx.Done` channel and will emit query cancellation
error when it sees that the context is canceled (so "query canceled"
error is also expected as part of the graceful shutdown). Previously, we
weren't swallowing it, and we do.

Addresses: #51647.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51772)
<!-- Reviewable:end -->

yuzefovich committed:
colflow: fix cancellation of the outbox

Previously, we were passing in the cancellation function of the flow's
context to each outbox which would result in too early of a cancellation
whenever any outboxes encounters a stream error. This is now fixed by
deriving a separate child context with separate cancellation function
for each outbox and only canceling the flow's context when the last
outbox exits.

Release note (bug fix): CockroachDB could previously encounter benign
internal "context canceled" errors when queries were executed by the
vectorized engine.

yuzefovich commented:
Ok, I think I have a pretty good understanding of the situation. I'm quite optimistic that this PR will resolve the "query canceled" and "context canceled" errors we have been seeing.

yuzefovich committed:
colexec: swallow "query canceled" in synchronizer with internal cancellation

Parallel unordered synchronizer shuts down all of its input subtrees by
canceling the context which those inputs are listening on. This results
in an "internal cancellation", and the error handling is different in
such case in the synchronizer - it needs to swallow "context canceled"
errors because, well, it was the one that canceled it, so such error is
expected part of the "graceful" shutdown. However, we also need to
swallow "query canceled" errors because `CancelChecker` operator might
be listening on `ctx.Done` channel and will emit query cancellation
error when it sees that the context is canceled (so "query canceled"
error is also expected as part of the graceful shutdown). Previously, we
weren't swallowing it, and we do.

Release note: None

knz commented on pkg/sql/colexec/parallel_unordered_synchronizer.go:
```diff
@@ -220,8 +221,13 @@ func (s *ParallelUnorderedSynchronizer) init(ctx context.Context) {
 			sendErr := func(err error) {
-				if strings.Contains(err.Error(), context.Canceled.Error()) && atomic.LoadInt32(&internalCancellation) == 1 {
-					// Don't propagate an internal cancellation error.
+				ctxCanceled := strings.Contains(err.Error(), context.Canceled.Error())
```
NB: These checks are still unsafe. Can't you use `errors.Is` here?

If not please file an issue to properly serialize errors and add a TODO here.

jordanlewis reviewed (approved):
LGTM aside from Raphael's comment.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51772)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51772#-MCs1oC0ID-i91ryV-9B-r2-224:-MDFxrTPDFe_Ep-vphit:b9zwkzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8a5c92abcd17a90907fbda299b8777a90fb151/pkg/sql/colexec/parallel_unordered_synchronizer.go#L224)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

NB: These checks are still unsafe. Can't you use `errors.Is` here?

If not please file an issue to properly serialize errors and add a TODO here.
</blockquote></details>

We can (for "pure" context cancellation errors and for query cancellation ones). I added a helper method to check for gRPC context cancellation (previously, it was handled by string matching for `context canceled`).


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
